### PR TITLE
feat(RP2040): implemented RESET_BASE peripherals

### PR DESCRIPTION
**Pull request by Turro75**

### CODE REVIEW

**Comment by Turro75** [comment]
It does nothing special, it just remove some annoying unimplemented peripherals output messages

**Comment by urish** [comment]
> It does nothing special, it just remove some annoying unimplemented peripherals output messages.

Clearing the noise is very important!

File: src/peripherals/reset.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import { BasePeripheral, Peripheral } from './peripheral';
+
+const RESET = 0x0; //Reset control.
+const WDSEL = 0x4; //Watchdog select.
+const RESET_DONE = 0x8; //Reset Done
+
+export class RPReset extends BasePeripheral implements Peripheral {
+  private reset: number = 0;
```
[the remark this answers is not in the document]
Comment (Turro75) [reply src/peripherals/reset.ts:8]: will You implement reset/watchdog handling? under normal operation this should be 0 that means the reset is not in progress.
mbed and picosdk does not seem writing to it , just reading. Anyway setting it to 0x1ffffff seems having zero impact.

**Comment by urish** [comment]
Thanks!

Can you please move the hardware divider code into a different pull request? 

One way to do it is to create a new branch from master, e.g. hardware-divider:

```
git fetch origin
git checkout -b hardware-divider origin/master
git cherry-pick 52d0c3d559aa116af2d2d3260b297ebbbe757844
```

then create a pull request from that branch (the last command copies a specific commit into the new branch)

Also, can you please check the new divider implementation against [hello_divider](https://github.com/raspberrypi/pico-examples/tree/master/divider) and [pico_divider_test](https://github.com/raspberrypi/pico-examples/tree/master/divider) (@kilograham mentioned that the version in `develop` branch is the most recent one)?

File: src/peripherals/reset.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (urish) [reply src/peripherals/reset.ts:8]: I guess we'll have watchdog at some point in the distance future.

Anyway, I just checked with GDB, and it seems like the datasheet is not precisely correct:

```
(gdb) monitor reset halt
target halted due to debug-request, current mode: Thread
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
target halted due to debug-request, current mode: Thread
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
(gdb) x/x 0x4000c000
0x4000c000:     0x00000000
```

It's 0 on reset (at least when debugging through GDB). So no need to change!

File: src/sio.ts [hunk 2]
```diff
@@ -117,6 +155,30 @@ export class RPSIO {
...
+      case DIV_UDIVIDEND:
+        this.divUDividend = value >>> 0;
+        this.divCSR = 1;
+        this.divQuotient = Math.trunc(this.divUDividend / this.divUDivisor);
+        this.divRemainder = this.divUDividend % this.divUDivisor;
+        break;
+      case DIV_SDIVIDEND:
+        this.divSDividend = value;
+        this.divCSR = 1;
+        this.divQuotient = Math.trunc(this.divSDividend % this.divSDivisor);
```
Comment (kilograham) [inline src/sio.ts:167]: should be /

File: src/sio.ts [hunk 3]
```diff
@@ -117,6 +155,30 @@ export class RPSIO {
...
+      case DIV_UDIVISOR:
+        this.divUDivisor = value >>> 0;
+        this.divCSR = 1;
+        this.divQuotient = Math.trunc(this.divUDividend / this.divUDivisor);
+        this.divRemainder = this.divUDividend % this.divUDivisor;
+        break;
+      case DIV_SDIVISOR:
+        this.divSDivisor = value;
+        this.divCSR = 1;
+        this.divQuotient = Math.trunc(this.divSDividend % this.divSDivisor);
```
Comment (kilograham) [inline src/sio.ts:179]: should be /

**Review by kilograham — COMMENTED** [review]
> Also, can you please check the new divider implementation against [hello_divider](https://github.com/raspberrypi/pico-examples/tree/master/divider) and [pico_divider_test](https://github.com/raspberrypi/pico-examples/tree/master/divider) (@kilograham mentioned that the version in `develop` branch is the most recent one)?

note that the implementation is incorrect (although it is possible the data sheet is not clear). there are not separate unsigned/signed divisor/dividended registers. writing to SDIVIDEND also writes to UDIVIDEND. Which register (signed or unsigned) was written last affects the sign of the result. Note also, that you can write to the result registers. These subtleties will affect save/restore of divider state (either if used in IRQs, or a RTOS) (also see `pico_divider_nesting_test`)

**Pull request closed without merging**